### DISCUSSION TITLE
Export isRedirectError/isNotFoundError

### DISCUSSION THREAD

**Discussion by Cretezy** [vercel/next.js#50170] [Ideas]
### Goals

Currently, it's impossible to catch errors from Server Actions as `redirect`s/`not-found`s are also thrown as error, which means they are caught if a Server Action (either on the server or client) is caught).

Example 1:

```ts
"use server"

import { isRedirectError } from "next/dist/client/components/redirect";

export async function serverAction() {
  try {
    doStuff();

    redirect("/");
  } catch (error) {
    // This will catch the redirect. Must use isRedirectError/isNotFoundError and rethrow
    if (isRedirectError(error) || isNotFoundError(error)) {
      throw error;
    }

    // Normal handling for doStuff()
  }
}
```

Example 2:

```ts
"use server"

export async function serverAction() {
  doStuff();

  redirect("/");
}



"use client";

import { isRedirectError } from "next/dist/client/components/redirect";

export function ClientComponent() {
  const [isPending, startTransition] = useTransition();

  function handleClick() {
    startTransition(async () => {
      try {
        await serverAction()
      } catch (error) {
        // This will catch the redirect. Must use isRedirectError/isNotFoundError and rethrow
        if (isRedirectError(error) || isNotFoundError(error)) {
          throw error;
        }
    
        // Normal handling for doStuff()
      }
    });
  }

  return <button onClick={handleClick}>Run</button>;
}
```

### Background

Current alternative is to import `isRedirectError`/`isNotFoundError` from the dist bundle and check manually.

### Proposal

3 items:

1. Export `isRedirectError`/`isNotFoundError` (and others like it, e.g. revalidate) from `next/navigation`
2. Create a `isServerActionError` (or `isNextNavigationError`), which checks for `isRedirectError`/`isNotFoundError` and all others like it

---

3. Possibly create a `handleServerActionError`, which can be added to a `catch`, which rethrows any navigation errors. Example code:
  ```ts
  function handleServerActionError(error) {
    if (isServerActionError(error)) throw error;
  }

  // Example usage:
  "use server"
  export async function serverAction() {
    try {
      doStuff();
  
      redirect("/");
    } catch (error) {
      handleServerActionError(error);
  
      // Normal handling for doStuff()
    }
  }
  ```

**Comment by codinginflow** [comment]
Maybe that's why Dan Abramov recommends to return the error in the body of the server action?

https://twitter.com/dan_abramov/status/1654894995315253257

I've been wondering what is wrong about just try/catching them in the client component.

**Reply by Cretezy** [reply]
Yeah this is what I've been doing so far. Especially since Errors are serialized over the wire.

It would still be nice to have these exports to do better handling, especially for wrapping libraries.

**Reply by eric-burel** [reply]
`isRedirectError` is not exported this way anymore, what's the current way of accessing it in 15.3 ?

**Reply by doooooooooots** [reply]
import { isRedirectError } from "next/dist/client/components/redirect-error";

**Reply by darthmaim** [reply]
No, you should use [`rethrow`](https://nextjs.org/docs/app/api-reference/functions/unstable_rethrow).
